pkg/client/cache/controller.go

```diff
@@ -61,21 +61,23 @@ type Config struct {
 type ProcessFunc func(obj interface{}) error
 
 // Controller is a generic controller framework.
-type Controller struct {
+type controller struct {
 	config         Config
 	reflector      *Reflector
 	reflectorMutex sync.RWMutex
 }
 
 // TODO make the "Controller" private, and convert all references to use ControllerInterface instead
-type ControllerInterface interface {
+type Controller interface {
 	Run(stopCh <-chan struct{})
 	HasSynced() bool
+	LastSyncResourceVersion() string
+	Requeue(obj interface{}) error
```

Comment:
> @deads2k, What's naughty about it? We don't take its ownership, we just need to inject an object there.

It means that you'll never be able to use this interface like all other consumers.  The controller is used by all other controllers to receive and this controller is using it to produce.  Used this way you can never participate in shared watches or shared caches.

@k82cn It's not a direction I want to encourage by placing in an interface (this is the only black sheep today).  You'd need some alternate way of satisfying the PV controller.



> On the other hand, PV controller is already pretty complicated enough and this would add another layer of complexity.

Is this the controller with the "here there be dragons, trust us, its good" comment?